src/engine/DepthOrderedCollection.hpp

```diff
@@ -0,0 +1,101 @@
+#pragma once
+#include <iterator>
+#include <map>
+#include <memory>
+#include <set>
+#include <vector>
+
+#include "IOrderedObject.hpp"
+
+namespace admirals {
+
+template <typename T = IOrderedObject> class DepthOrderedCollection {
+private:
+    // Should be sorted with respect to game object index, lower first.
+    struct Comparator {
+        Comparator(const DepthOrderedCollection &c) : m_collection(c) {}
+
+        bool operator()(const std::string &l, const std::string &r) const {
+            const std::shared_ptr<IOrderedObject> a =
+                m_collection.m_objects.at(l);
+            const std::shared_ptr<IOrderedObject> b =
+                m_collection.m_objects.at(r);
+            return a->m_order < b->m_order;
+        }
+
+    private:
+        const DepthOrderedCollection &m_collection;
+    };
+
+    typedef std::map<std::string, std::shared_ptr<T>> MAP;
+    typedef std::multiset<std::string, Comparator> SET;
```

Comment:
These are not nice typenames. Perhaps something like `StringToSharedPtrMap` or something really descriptive.